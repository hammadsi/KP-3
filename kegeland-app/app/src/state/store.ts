import {combineReducers, configureStore} from '@reduxjs/toolkit';
import AsyncStorage from '@react-native-async-storage/async-storage';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import reduxFlipper from 'redux-flipper';

import appReducer from './ducks/app/app.reducer';
import bluetoothReducer from './ducks/bluetooth/bluetooth.reducer';
import authReducer from './ducks/auth/auth.reducer';
import questionsReducer from './ducks/questions/questions.reducer';
import sessionReducer from './ducks/session/session.reducer';

import rootSaga, {sagaMiddleware} from './rootSaga';

/**
 * Configuration for redux-persist.
 * Redux-persist is responsible for maintaining the state
 * between sessions.
 */
const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage,
  whitelist: ['app', 'auth', 'bluetooth'],
};

// Initiate the root reducer
export const rootReducer = combineReducers({
  app: appReducer,
  auth: authReducer,
  bluetooth: bluetoothReducer,
  questions: questionsReducer,
  session: sessionReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Check if development-environment is enabled
const isDev = !['production', 'test', 'integration'].includes(
  process.env.NODE_ENV || '',
);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        })
          .concat(logger)
          .concat(reduxFlipper())
          .concat(sagaMiddleware)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(sagaMiddleware),
  devTools: isDev,
});

// Start redux-saga
sagaMiddleware.run(rootSaga);

// Create persistor for persisting state
export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
