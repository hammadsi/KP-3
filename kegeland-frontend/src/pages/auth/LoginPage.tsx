import {
  Box,
  ButtonGroup,
  Center,
  Container,
  Heading,
  Link,
  Text,
  VStack,
} from '@chakra-ui/react';
import { Formik } from 'formik';
import { InputControl, SubmitButton } from 'formik-chakra-ui';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import * as Yup from 'yup';

import useAppDispatch from '../../hooks/useAppDispatch';
import useAppSelector from '../../hooks/useAppSelector';
import { signInUser } from '../../state/ducks/auth/auth.actions';
import { LoginDTO } from '../../state/ducks/auth/auth.interface';
import { clearError } from '../../state/ducks/auth/auth.reducer';

const validateSchema = Yup.object({
  email: Yup.string().required().email('Email is not valid').label('Email'),
  password: Yup.string().required().label('Password'),
});

const styles = {
  button: {
    margin: '0 0rem',
    padding: '0.5rem 1rem',
    border: 'none',
    fontSize: '20px',
    fontWeight: 'bold',
    minWidth: '150px',
    borderBottom: '1px solid black',
  },
  buttonGroup: {
    borderBottom: '1px solid black',
  },
};

const LoginPage = () => {
  const navigate = useNavigate();
  const { error, loading } = useAppSelector((state) => state.auth);
  const dispatch = useAppDispatch();
  const { isSignedIn } = useAppSelector((state) => state.auth);
  const [isPatient, setIsPatient] = useState(true);

  const handleChange = (boolean: boolean) => {
    setIsPatient(boolean);
  };

  const signIn = (data: LoginDTO) => {
    dispatch(signInUser(data));
  };

  useEffect(() => {
    dispatch(clearError());
    if (isSignedIn) {
      navigate('/');
    }
  }, [navigate, isSignedIn]);

  return (
    <Center width="100%" height="100vh">
      <Container>
        <Container paddingTop="1em">
          <Formik
            onSubmit={async (values) => {
              signIn(values);
            }}
            initialValues={{ email: '', password: '' }}
            validationSchema={validateSchema}>
            {(formProps) => (
              <Box
                borderWidth="1px"
                backgroundColor="white"
                rounded="lg"
                shadow="1px 1px 3px rgba(0,0,0,0.3)"
                maxWidth={800}
                p={6}
                m="10px auto"
                as="form"
                onSubmit={(e: any) => {
                  e.preventDefault();
                  formProps.handleSubmit();
                }}>
                <VStack spacing={5} align="stretch">
                  <Box>
                    <Heading as="h3" size="lg" textAlign="center">
                      Log in{' '}
                    </Heading>
                  </Box>
                  <Box textAlign="center">
                    <ButtonGroup>
                      <button
                        value="Patient"
                        style={{
                          ...styles.button,
                          color: isPatient ? '#03a9f4' : 'black',
                          borderBottom: isPatient
                            ? '3px solid #03a9f4'
                            : 'none',
                        }}
                        onClick={() => handleChange(true)}>
                        Patient
                      </button>
                      <button
                        value="Physician"
                        style={{
                          ...styles.button,
                          color: !isPatient ? '#03a9f4' : 'black',
                          borderBottom: !isPatient
                            ? '3px solid #03a9f4'
                            : 'none',
                        }}
                        onClick={() => handleChange(false)}>
                        Physician
                      </button>
                    </ButtonGroup>
                  </Box>
                  <Box>
                    <InputControl
                      inputProps={{
                        type: 'email',
                        placeholder: 'ola.nordmann@example.com',
                      }}
                      name="email"
                      label="Email address"
                      data-testid="email-input"
                    />
                  </Box>
                  <Box>
                    <InputControl
                      name="password"
                      label="Password"
                      data-testid="password-input"
                      inputProps={{
                        type: 'password',
                        autoComplete: 'password',
                        placeholder: '• • • • • • • •',
                      }}
                    />
                  </Box>
                  <Box color={'red'}>{error}</Box>
                  <div
                    style={{
                      display: 'flex',
                      justifyContent: 'space-between',
                    }}>
                    {isPatient ? (
                      <Box textAlign="left">
                        <Text color="#black">
                          Not a member yet?
                          <Link color="primary.600" href="/register">
                            <b> Register!</b>
                          </Link>
                        </Text>
                      </Box>
                    ) : (
                      <h1></h1>
                    )}
                    <Box textAlign="right">
                      <SubmitButton
                        colorScheme="primary"
                        isLoading={formProps.isSubmitting || loading}
                        isDisabled={!formProps.isValid}>
                        Log in
                      </SubmitButton>
                    </Box>
                  </div>
                </VStack>
              </Box>
            )}
          </Formik>
        </Container>
      </Container>
    </Center>
  );
};

export default LoginPage;
