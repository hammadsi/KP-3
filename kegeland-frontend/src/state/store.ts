import { combineReducers, configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import authReducer from './ducks/auth/auth.reducer';
import patientsReducer from './ducks/patients/patients.reducer';
import questionnairesReducer from './ducks/questionnaires/questionnaires.reducer';
import { sensorReducer } from './ducks/sensors/sensors.reducer';
import sessionsReducer from './ducks/sessions/sessions.reducer';
import settingsReducer from './ducks/settings/settings.reducer';
import wheelchairPatientsReducer from './ducks/wheelchairPatients/wheelchairPatients.reducer';

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
  whitelist: ['settings', 'auth'],
};

export const rootReducer = combineReducers({
  settings: settingsReducer,
  auth: authReducer,
  patients: patientsReducer,
  sensors: sensorReducer,
  sessions: sessionsReducer,
  questionnaires: questionnairesReducer,
  wheelchairPatients: wheelchairPatientsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const isDev = !['production', 'test'].includes(process.env.NODE_ENV || '');

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(logger)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }),
  devTools: isDev,
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
