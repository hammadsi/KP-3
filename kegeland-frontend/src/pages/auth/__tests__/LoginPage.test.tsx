import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import { mockStore } from '../../../state/mocks/store.mock';
import LoginPage from '../LoginPage';

const mockedNavigate = jest.fn();

jest.mock('react-router-dom', () => {
  const actualNav = jest.requireActual('react-router-dom');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('Test excercise', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={mockStore()}>
          <MemoryRouter>
            <LoginPage />
          </MemoryRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
