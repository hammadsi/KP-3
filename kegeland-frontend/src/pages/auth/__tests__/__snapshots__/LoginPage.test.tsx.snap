// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test excercise renders correctly 1`] = `
<div
  className="css-lt6aye"
>
  <div
    className="chakra-container css-0"
  >
    <div
      className="chakra-container css-1c2bf3b"
    >
      <form
        className="css-omnk1d"
        onSubmit={[Function]}
      >
        <div
          className="chakra-stack css-1ws1qxv"
        >
          <div
            className="css-0"
          >
            <h3
              className="chakra-heading css-xi606m"
            >
              Log in
               
            </h3>
          </div>
          <div
            className="css-0"
          >
            <h3
              className="chakra-heading css-xi606m"
            >
              Please enter email and password to log in.
            </h3>
          </div>
          <div
            className="css-0"
          >
            <div
              className="chakra-form-control css-0"
              data-testid="email-input"
              role="group"
            >
              <label
                className="chakra-form__label css-uvho5g"
                htmlFor="email"
                id="field-:r0:-label"
              >
                Email address
                <span
                  aria-hidden={true}
                  className="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <input
                aria-required={true}
                className="chakra-input css-0"
                disabled={false}
                id="email"
                name="email"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="ola.nordmann@example.com"
                readOnly={false}
                required={true}
                type="email"
                value=""
              />
            </div>
          </div>
          <div
            className="css-0"
          >
            <div
              className="chakra-form-control css-0"
              data-testid="password-input"
              role="group"
            >
              <label
                className="chakra-form__label css-uvho5g"
                htmlFor="password"
                id="field-:r1:-label"
              >
                Password
                <span
                  aria-hidden={true}
                  className="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <input
                aria-required={true}
                autoComplete="password"
                className="chakra-input css-0"
                disabled={false}
                id="password"
                name="password"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="• • • • • • • •"
                readOnly={false}
                required={true}
                type="password"
                value=""
              />
            </div>
          </div>
          <div
            className="css-tokvmb"
          />
          <div
            className="css-s2uf1z"
          >
            <button
              className="chakra-button css-4xx2wk"
              disabled={false}
              type="submit"
            >
              Log in
            </button>
          </div>
          <div
            className="css-13brihr"
          >
            <p
              className="chakra-text css-1lve7oc"
            >
              Not a member yet?
              <a
                className="chakra-link css-lalzaa"
                href="/register"
              >
                <b>
                   Register!
                </b>
              </a>
            </p>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
